src/core/include/units/quantity_spec.h

```diff
@@ -0,0 +1,518 @@
+ * For example:
+ *
+ * @code{.cpp}
+ * // quantity kinds
+ * inline constexpr struct width : quantity_spec<length> {} width;
+ * inline constexpr struct height : quantity_spec<length> {} height;
+ * inline constexpr struct thickness : quantity_spec<width> {} thickness;
+ * inline constexpr struct diameter : quantity_spec<width> {} diameter;
+ * inline constexpr struct position_vector : quantity_spec<length, quantity_character::vector> {} position_vector;
+ *
+ * // derived quantities
+ * inline constexpr struct area : quantity_spec<pow<2>(length)> {} area;
+ * inline constexpr struct volume : quantity_spec<pow<3>(length)> {} volume;
+ * inline constexpr struct velocity : quantity_spec<position_vector / duration> {} velocity;  // vector
+ * inline constexpr struct speed : quantity_spec<distance / duration> {} speed;
+ * inline constexpr struct force : quantity_spec<mass * acceleration, quantity_character::vector> {} force;
+ * inline constexpr struct power : quantity_spec<force * velocity, quantity_character::scalar> {} power;
+ * @endcode
+ *
+ * Two quantity specifications are deemed equal when they are of the same type. With that, both strong
+ * types `speed` and `velocity` are considered not equal to `derived_dimension<length, per<time>>` or
+ * to each other. However, they are both convertible to `derived_dimension<length, per<time>>`.
+ * User can implicitly convert up and down the inheritance hierarchy between them. However, `speed` and
+ * `velocity` are not convertible as they do not inherit from each other. They are from two separate branches
+ * of dimensionally equivalent quantities.
+ *
+ * @tparam Q quantity specification of a parent quantity
+ * @tparam Args optionally a value of a `quantity_character` in case the base quantity should not be scalar
+ */
+#ifdef __cpp_explicit_this_parameter
+template<QuantitySpec auto Q, one_of<quantity_character> auto... Args>
+struct quantity_spec<Q, Args...> : std::remove_const_t<decltype(Q)> {
+#else
+template<typename Self, QuantitySpec auto Q, one_of<quantity_character> auto... Args>
+struct quantity_spec<Self, Q, Args...> : std::remove_const_t<decltype(Q)> {
+#endif
+  static constexpr auto kind_of = Q;
+  static constexpr quantity_character character = detail::quantity_character_init<Args...>(Q.character);
+
+#ifndef __cpp_explicit_this_parameter
+  template<Unit U>
+  [[nodiscard]] constexpr auto operator[](U u) const
+    requires(this->dimension == detail::get_dimension_for(u))
+  {
+    return reference<Self{}, u>{};
+  }
+#endif
+};
+
+#ifdef __cpp_explicit_this_parameter
+
+#define QUANTITY_SPEC(name, ...)                                \
+  inline constexpr struct name : quantity_spec<##__VA_ARGS__> { \
+  } name
+
+#else
+
+#define QUANTITY_SPEC(name, ...)                                      \
+  inline constexpr struct name : quantity_spec<name, ##__VA_ARGS__> { \
+  } name
+
+#endif
+
+// TODO Should a quantity_cast be allowed to cast between speed and velocity?
```

Comment:
We also have to take representation types into account here. For some use cases `int` can be both a vector and scalar character in such case those probably could be interconvertible (?). For other cases where `int` is not a vector then I agree it could be convertible in one direction only which may require a lot of code changes as for now everything works on an `inconvertible()`.

Reply:
Also, how to get a magnitude of a vector in a generic way? Should we introduce some customization point for that?

Reply:
> Also, how to get a magnitude of a vector in a generic way? Should we introduce some customization point for that?

Depends on the scope of the library. I'd certainly do it for my code, if the need arises.

> For some use cases `int` can be both a vector and scalar character in such case those probably could be interconvertible (?).

Like a vector of 1 dimension?

Reply:
Yes, ISO 80000-2 explicitly states:
> A vector is a tensor of the first order and a scalar is a tensor of order zero

Reply:
Yes. That also ties to the "force" portion of https://github.com/mpusz/units/issues/405#issuecomment-1339547543

Reply:
Another related issue #409.